presto-main/src/main/java/com/facebook/presto/matching/MatchingEngine.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.matching;
+
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Streams;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class MatchingEngine<T extends Matchable>
+{
+    private final ListMultimap<Class, T> matchablesByClass;
+
+    private MatchingEngine(ListMultimap<Class, T> matchablesByClass)
+    {
+        this.matchablesByClass = ImmutableListMultimap.copyOf(matchablesByClass);
+    }
+
+    public Stream<T> getCandidates(Object object)
```

Comment:
I'm hesitant whether this class and the `Matchable` interface really belong to the `matching` package. The only public method here is `getCandidates`, which is at most some form of `pre-matching`. I see two ways out of this:
1. leave this in the `planner` package and call it `RuleIndex` - then the `Matchable` interface and type parameter are not needed
2. make this an additional entry point to the matching features and call it `PreMatcher`.

I wouldn't call it a `MatchingEngine`, as:
1. it only does pre-matching
2. the actual matching logic is within the matchers.